src/Build/Collections/PropertyDictionary.cs

```diff
@@ -42,6 +44,11 @@ internal sealed class PropertyDictionary<T> : IEnumerable<T>, IEquatable<Propert
         [DebuggerBrowsable(DebuggerBrowsableState.RootHidden)]
         private readonly RetrievableEntryHashSet<T> _properties;
 
+        /// <summary>
+        /// Lock object to guard access to backing collection.
+        /// </summary>
+        private readonly ReaderWriterLockSlim _lock = new ReaderWriterLockSlim();
```

Comment:
Would be nice to reuse the `DisposableReaderWriterLockSlim`, and maybe make it its own file.

Reply:
In [https://github.com/cdmihai/msbuild/blob/c8660645471a185287148f9e0897a3ec9c1ec3c8/src/Build/Definition/ProjectCollection.cs#L87](https://github.com/cdmihai/msbuild/blob/c8660645471a185287148f9e0897a3ec9c1ec3c8/src/Build/Definition/ProjectCollection.cs#L87)

---
In reply to: [302275736](https://github.com/microsoft/msbuild/pull/4374#discussion_r302275736) [](ancestors = 302275736)